clusterctl/clusterdeployer/clusterdeployer.go

```diff
@@ -148,39 +151,50 @@ func (d *ClusterDeployer) Create(cluster *clusterv1.Cluster, machines []*cluster
-	if err != nil {
-		return fmt.Errorf("unable to create internal cluster: %v", err)
-	}
-	defer closeClient(internalClient, "internal")
+	targetClient := externalClient
+	targetCluster := "external"
 
-	glog.Info("Applying Cluster API stack to internal cluster")
-	if err := d.applyClusterAPIStackWithPivoting(internalClient, externalClient); err != nil {
-		return fmt.Errorf("unable to apply cluster api stack to internal cluster: %v", err)
-	}
+	if d.pivotCluster {
+		glog.Info("Creating internal cluster")
+		internalClient, err := d.createInternalCluster(externalClient, provider, kubeconfigOutput)
+		if err != nil {
+			return fmt.Errorf("unable to create internal cluster: %v", err)
+		}
+		defer closeClient(internalClient, "internal")
+		targetClient = internalClient
+		targetCluster = "internal"
 
-	glog.Info("Saving provider components to the internal cluster")
-	err = d.saveProviderComponentsToCluster(providerComponentsStoreFactory, kubeconfigOutput)
-	if err != nil {
-		return fmt.Errorf("unable to save provider components to internal cluster: %v", err)
-	}
+		glog.Info("Applying Cluster API stack to internal cluster")
+		if err := d.applyClusterAPIStackWithPivoting(internalClient, externalClient); err != nil {
+			return fmt.Errorf("unable to apply cluster api stack to internal cluster: %v", err)
+		}
+
+		glog.Info("Saving provider components to the internal cluster")
+		err = d.saveProviderComponentsToCluster(providerComponentsStoreFactory, kubeconfigOutput)
+		if err != nil {
+			return fmt.Errorf("unable to save provider components to internal cluster: %v", err)
+		}
 
-	// For some reason, endpoint doesn't get updated in external cluster sometimes. So we
-	// update the internal cluster endpoint as well to be sure.
-	glog.Infof("Updating internal cluster object with master (%s) endpoint", master.Name)
-	if err := d.updateClusterEndpoint(internalClient, provider); err != nil {
-		return fmt.Errorf("unable to update internal cluster endpoint: %v", err)
+		// For some reason, endpoint doesn't get updated in external cluster sometimes. So we
+		// update the internal cluster endpoint as well to be sure.
+		glog.Infof("Updating internal cluster object with master (%s) endpoint", master.Name)
+		if err := d.updateClusterEndpoint(internalClient, provider); err != nil {
+			return fmt.Errorf("unable to update internal cluster endpoint: %v", err)
+		}
+	} else {
+		if _, err := d.retrieveKubeConfig(externalClient, provider, kubeconfigOutput); err != nil {
+			return fmt.Errorf("unable to retrieve the provisioned cluster kubeconfig: %v", err)
+		}
 	}
 
-	glog.Info("Creating node machines in internal cluster.")
-	if err := internalClient.CreateMachineObjects(nodes); err != nil {
+	glog.Infof("Creating node machines in %s cluster.", targetCluster)
+	if err := targetClient.CreateMachineObjects(nodes); err != nil {
 		return fmt.Errorf("unable to create node machines: %v", err)
 	}
 
 	if d.addonComponents != "" {
```

Comment:
What is purpose of the addon components? Do they extend the cluster API stack? Checking:
* https://github.com/kubernetes-sigs/cluster-api-provider-gcp/blob/master/clusterctl/examples/google/addons.yaml.template
* https://github.com/roberthbailey/cluster-api-provider-vsphere/blob/master/clusterctl/examples/vsphere/addons.yaml.template

I can't decide if they are meant for the external of the internal cluster.

Reply:
They are meant for the internal cluster. The provider components are the cluster api stack that is applied to the internal and external cluster. The addon components are things we "add on" to the internal cluster in addition to the core k8s control plane as part of the baseline installation.